File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/javascript/08.Combination Sum 2.js
/*
=============================================================================
  QUESTION: 40. Combination Sum II (LeetCode)
=============================================================================

  Given candidates (MAY HAVE DUPLICATES) and target, find all unique
  combinations summing to target. Each number used ONLY ONCE.

  Example: [10,1,2,7,6,1,5], target=8 → [[1,1,6],[1,2,5],[1,7],[2,6]]

=============================================================================
  APPROACH: Pick / Not-Pick + Skip Duplicates — O(N × 2^N) Time, O(target) Space
=============================================================================

  Combines: Combination Sum (target tracking) + Subset 2 (skip duplicates).

  vs Combination Sum 1:
  - Each element used ONCE → move to index+1 when picking
  - Has duplicates → sort + skip all duplicates when excluding

  Dry Run sorted [1,1,2,5,6,7,10], target=8:
    pick 1(idx0) → pick 1(idx1) → pick 6 → t=0 → PUSH [1,1,6] ✓
                 → skip 1(idx1) → pick 2 → pick 5 → t=0 → PUSH [1,2,5] ✓
                                → skip 2 → pick 7 → t=0 → PUSH [1,7] ✓
    skip 1(idx0) → skip 1(idx1) too → pick 2 → pick 6 → t=0 → PUSH [2,6] ✓
    Result: [[1,1,6],[1,2,5],[1,7],[2,6]] ✓

=============================================================================
*/

function combinationSum2(candidates, target) {
    candidates.sort((a, b) => a - b);
    let result = [];

    function solve(index, target, temp) {
        if (target === 0) { result.push([...temp]); return; }
        if (index === candidates.length || target < 0) return;

        // PICK current element (move to next — used once)
        if (candidates[index] <= target) {
            temp.push(candidates[index]);
            solve(index + 1, target - candidates[index], temp);
            temp.pop();
        }

        // Skip all duplicates, then EXCLUDE
        while (index + 1 < candidates.length && candidates[index] === candidates[index + 1])
            index++;
        solve(index + 1, target, temp);
    }

    solve(0, target, []);
    return result;
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(combinationSum2([10, 1, 2, 7, 6, 1, 5], 8));
// [[1,1,6],[1,2,5],[1,7],[2,6]]

File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/python/08.Combination Sum 2.py
"""
=============================================================================
  QUESTION: 40. Combination Sum II (LeetCode)
=============================================================================

  MAY HAVE DUPLICATES, each used ONLY ONCE. No duplicate combinations.
  Example: [10,1,2,7,6,1,5], target=8 → [[1,1,6],[1,2,5],[1,7],[2,6]]

  APPROACH: Combination Sum + Subset 2 duplicate-skip logic
  - Pick: move to index+1 (each used once)
  - Not-pick: skip all duplicates, then move to next

  DRY RUN sorted [1,1,2,5,6,7,10], target=8:
    Pick 1(0) → Pick 1(1) → ... → Pick 6 → target=0 → [1,1,6] ✓
    Pick 1(0) → Skip 1(1) → Pick 2 → Pick 5 → target=0 → [1,2,5] ✓
    Pick 1(0) → ... → Pick 7 → target=0 → [1,7] ✓
    Skip both 1s → Pick 2 → Pick 6 → target=0 → [2,6] ✓

  Time: O(N × 2^N), Space: O(target)
=============================================================================
"""


def solve(index, target, candidates, temp, ans):
    if target == 0:
        ans.append(temp[:])
        return
    if index == len(candidates) or target < 0:
        return

    if candidates[index] <= target:
        temp.append(candidates[index])
        solve(index + 1, target - candidates[index], candidates, temp, ans)
        temp.pop()

    # Skip duplicates
    while index + 1 < len(candidates) and candidates[index] == candidates[index + 1]:
        index += 1

    solve(index + 1, target, candidates, temp, ans)


def combination_sum2(candidates, target):
    ans = []
    candidates.sort()
    solve(0, target, candidates, [], ans)
    return ans


# Driver Code
if __name__ == "__main__":
    print(combination_sum2([10, 1, 2, 7, 6, 1, 5], 8))
    # [[1,1,6],[1,2,5],[1,7],[2,6]]

